src/common/mongoose/plugins/history.plugin.ts

```diff
@@ -0,0 +1,338 @@
+  modelName?: string;
+  getActiveUser?: () => string | undefined; // Function to get current user context if needed
+  configService?: ConfigService; // Optional ConfigService for accessing environment variables
+  // Add these two new options
+  trackableStrategy?: "delta" | "document";
+  trackables?: string[];
+}
+
+// Define an interface for the query context including our custom property
+// Make it generic over the Document type (T) and the Result type (Res)
+// Use QueryWithHelpers for better compatibility with Mongoose's internal types
+interface QueryWithHistory<
+  Res, // Result type of the query (e.g., T | null, DeleteResult)
+  T extends Document, // Document type
+> extends QueryWithHelpers<Res, T, object> {
+  // Use object for THelpers
+  _originalDoc?: T | (T & Document<unknown, object, T>[]) | null; // Store the original doc(s) fetched
+}
+
+export function historyPlugin(
+  schema: Schema,
+  options: HistoryPluginOptions = {},
+) {
+  const {
+    historyModelName = GenericHistory.name,
+    modelName: optionsModelName, // Extract modelName if provided in options
+    getActiveUser: getActiveUser,
+    configService,
+    // Extract from options with fallbacks to ConfigService values
+    trackableStrategy: explicitStrategy,
+    trackables: explicitTrackables,
+  } = options;
+
+  // Use options if provided, otherwise try ConfigService, then fall back to defaults
+  const trackableStrategy =
+    explicitStrategy ??
+    (configService?.get<string>("trackableStrategy") === "delta"
+      ? "delta"
+      : "document");
+
+  const trackables =
+    explicitTrackables ?? configService?.get<string[]>("trackables") ?? [];
+
+  // Get the model name from options
+  if (!optionsModelName) {
+    // If not provided, warn and skip setup
+    console.warn(
+      "HistoryPlugin: Could not determine model name for schema. Please provide a modelName in plugin options.",
+    );
+    return; // Skip setup if we can't determine the model name
+  }
+
+  const modelName = optionsModelName;
+
+  // Skip setting up the plugin if the model is not trackable
+  if (!trackables.includes(modelName)) {
+    return;
+  }
+  console.log(`History tracking enabled for model: ${modelName}`);
+
+  // Middleware for update operations ('findOneAndUpdate')
+  schema.pre(
+    "findOneAndUpdate",
+    // Specify the expected result type (Document | null) and the Document type
+    async function (this: QueryWithHistory<Document | null, Document>, next) {
+      // Type 'this' to include _originalDoc
+      try {
+        const docToUpdate = await this.model
+          .findOne(this.getQuery())
+          .lean<Document>() // Keep lean<Document> or use a specific interface for the plain object
+          .exec();
+        if (docToUpdate) {
+          // Assign the potentially plain object from lean() to _originalDoc
+          // Cast lean result if needed, or adjust _originalDoc type
+          this._originalDoc = docToUpdate as Document;
+        }
```

Comment:
I wonder if we can consolidate all the calls to the database only in `schema.pre` or in `schema.post`?